Interface design for Multi-table query

Achilles-96 opened the pull request:
Not to merge.
I made layout for new UI. Please review this and let me know if I can proceed further implementing backend as well.
Issue - #13268 

Signed-off-by: Raghuram Vadapalli<raghuram.vadapalli@research.iiit.ac.in>

Before submitting pull request, please check that every commit:

- [x] Has proper Signed-Off-By
- [x] Has commit message which describes it
- [x] Is needed on it's own, if you have just minor fixes to previous commits, you can squash them
- [ ] Any new functionality is covered by tests

Achilles-96 committed:
Interface design for Multi-table query

Signed-off-by: Raghuram <raghuram.vadapalli@research.iiit.ac.in>

Achilles-96 commented:
Access the page by going to http://<pma-address>/db_multi_table_query.php?db=<db_name>
eg: http://localhost/pma-dev/db_multi_table_query.php?db=hello

ibennetch commented:
This looks like a good start to me. I don't have much feedback right now on this, I thought your mockup looked good and this is beginning to look really good as well. I think you're clearly on the right path.

Any opinions from other @phpmyadmin/developers ?

Achilles-96 commented:
This PR is ready for review.

ibennetch commented:
I'm currently getting an error:
> Warning in ./db_multi_table_query.php#12
require_once(libraries/sql.lib.php): failed to open stream: No such file or directory
> 
> Backtrace
> 
> ./db_multi_table_query.php#12: require_once()

Do you know what went wrong?

ibennetch commented:
I'm sure someone will correct me if I'm wrong, but I believe this is to replace the current Query tab and Visual Builder from db_designer. In that case, we probably should update the link from the Query tab to this new page and remove the old code (at least, db_qbe — I'm not certain offhand how much of the db_designer code is integrated with the rest of the Designer features).

Achilles-96 commented:
I am not getting that error. sql.lib.php is present in libraries. I don't understand why you are getting that error. 
I am trying to change the link for query tab.

Achilles-96 commented:
The link works now. Should I remove old code in this PR itself.

ibennetch commented:
I'll try again, if it's working for you then there must be a problem on my side. Thanks.

ibennetch commented:
I have a few thoughts about things that can be improved:

* After pressing Submit query, the results are shown under the query interface. I can scroll the results and down, but the query interface floats on top. I think ideally, it would collapse but still be present, so the user could go back and make changes; kind of like this:
![image](https://user-images.githubusercontent.com/1443710/27336661-4130a962-559e-11e7-99e1-b3c5ff261e32.png)
* When used with `$cfg['InitialSlidersState'] = 'open';`, the Criteria area is expanded by default (which is fine), but then clicking the checkbox to enable the criteria results in the area being hidden. I think the best solution to this is simply to remove the collapse/hide event from the checkbox — if a user wants to hide it, they can click the text "- criteria". For those users in the default state, clicking the checkbox will still expand the area, which is great.
* When I combine two tables with criteria fields, they seem to get concatenated in an odd way, such as
  > SELECT \`actor\`.\`actor_id\`, \`film\`.\`film_id\`
  > FROM \`actor\`, \`film\`
  > WHERE \`actor\`.\`actor_id\` = 1\`film\`.\`film_id\` = 1;

  ![image](https://user-images.githubusercontent.com/1443710/27337062-5c30d498-559f-11e7-8592-075c7d9e9240.png)

This looks nice, good work.

ibennetch reviewed (changes requested):
This doesn't quite work because libraries/Menu.php is still looking for the $tabs['qbe'] array but libraries/Util.php changes from qbe to 'multi_table_query'.

I have a slight preference toward renaming the array in Menu.php from qbe to multi_table_query.

Achilles-96 commented:
Regarding the first two points: They make sense and I will be on them.
Regarding combining two tables, you have to select whether to use `AND` or `OR` by selecting the radio button. I guess one of them should be checked by default.
I will also do changes in Menu.php as you mentioned.

Thanks for the review.

Achilles-96 commented:
Also, I want to suggest a different solution for slider thing. Can we add an additional parameter to slider to override default configuration. Please check my latest patch.
I feel the sliding the query window up after fetching result looks very good.

Achilles-96 commented:
Also, I put `checked` for `AND` by default in criteria. I don't understand why it is not checking. Except that, I believe I addressed all the concerns raised by you.

ibennetch commented on db_multi_table_query.php:
```diff
@@ -0,0 +1,59 @@
+if (isset($_REQUEST['sql_query'])) {
+    $sql_query = $_REQUEST['sql_query'];
+    $db = $_REQUEST['db'];
+    //    $response = Response::getInstance();
```
Is this line old or debugging code that can be removed?

Achilles-96 commented:
Hi @ibennetch , should I update the patch removing old code?

ibennetch commented:
Let's remove the old code as a separate pull request.

Achilles-96 commented:
What is the conflict here? I am able to merge it on my local repo.

ibennetch commented:
It's reporting that `libraries/Menu.php` and `libraries/Util.php` have conflicts, but when I tried to merge it manually, git was able to do so with a three-way merge. I think this means that Github is unable to merge automatically while maintaining proper history.

I believe the correct resolution is to rebase while in your branch; that will bring in the changes from master and allow the merge to complete successfully. It wouldn't hurt to make a copy of your branch first (`git checkout Multi-Table-Query-UI` then `git branch Multi-Table-backup` should do).

From within Multi-Table-Query-UI, you should be able to simply run `git rebase origin/master` and it will give a message like

> First, rewinding head to replay your work on top of it...

Then it should think for a few minutes and give you a rebased version of your current branch (with no commits lost, but with new commit IDs). Then you can push that back to your github fork which will update the pull request here.

I think that's safer than me trying to merge and rebase it myself.

Achilles-96 commented:
Moving files in `libraries/` caused some failures. I fixed them.

mauriciofauth commented on libraries/DbMultiTableQuery.php:
```diff
@@ -0,0 +1,188 @@
+ * @package PhpMyAdmin
+ */
+namespace PMA\libraries;
+use PhpMyAdmin;
```
I think it's best to specify which classes you're using instead of including the entire namespace.

mauriciofauth commented:
You can move the `DbMultiTableQuery` class to the `libraries/classes` directory and use the `PhpMyAdmin` namespace. This would save the work of moving this file in the future.

Achilles-96 commented:
Hi @mauriciofauth, thanks for the review. I updated the patch. Please verify.

mauriciofauth commented:
Hi, @Achilles-96. The only thing I see that is missing would be to move the `js/jquery.md5.js` file to `js/vendor/jquery.md5.js`, since it's a third-party library.

ibennetch commented on js/db_multi_table_query.js:
```diff
@@ -0,0 +1,230 @@
+AJAX.registerOnload('db_multi_table_query.js', function () {
+
+    var editor = PMA_getSQLEditor($('#MultiSqlquery'), {}, 'both');
+    $('.CodeMirror-line').css('text-align','left');
```
Can you add a space after the comma here? It will make Codacy happier and is more readable.

ibennetch commented on js/db_multi_table_query.js:
```diff
@@ -0,0 +1,230 @@
+        var table_aliases = {};
+        $('.tableNameSelect').each(function() {
+            $show = $(this).siblings('.show_col').first();
+            if ($(this).val() != "" && $show.prop('checked')) {
```
Can `!==` be used here instead of `!=`? The identity comparisons are generally considered safer than their equity counterparts (`===` instead of `==`).

ibennetch commented on js/db_multi_table_query.js:
```diff
@@ -0,0 +1,230 @@
+                var table_alias = $(this).siblings('.table_alias').first().val();
+                var column_alias = $(this).siblings('.col_alias').first().val();
+
+                if (table_alias != "") {
```
Here as well.

ibennetch commented on js/db_multi_table_query.js:
```diff
@@ -0,0 +1,230 @@
+                }
+            }
+        });
+        if (Object.keys(table_aliases).length == 0) {
```
And here.

ibennetch commented on js/db_multi_table_query.js:
```diff
@@ -0,0 +1,230 @@
+
+        query = 'SELECT ';
+        query += '`' + columns[0][0] + '`.`' + columns[0][1] + '`';
+        if (columns[0][2] != "") {
```
Another one.

ibennetch commented on js/db_multi_table_query.js:
```diff
@@ -0,0 +1,230 @@
+        }
+        for (var i = 1; i < columns.length; i++) {
+            query += ', `' + columns[i][0] + '`.`' + columns[i][1] + '`';
+            if (columns[i][2] != "") {
```
Here also.

ibennetch commented on js/db_multi_table_query.js:
```diff
@@ -0,0 +1,230 @@
+            query += '\nWHERE ';
+
+            var tables = [];
+            var columns = [];
```
I think you initialize the variables `tables` and `columns` here but I don't see them used; is this old code?

ibennetch reviewed (changes requested):
Okay, I think we're ready for merge here except for a few issues that Codacy found. I've commented on them in-line here; if you could take a look at those when you have a chance and let me know what you think.

Achilles-96 commented:
Any more issues with this?

nijel commented on libraries/classes/DbMultiTableQuery.php:
```diff
@@ -0,0 +1,187 @@
+
+        $html_output .= '</div>';
+        $html_output .= '</div>';
+        $html_output .= '<a href="#" title="Remove this column" style="float:right;color:red" class="removeColumn">X</a>';
```
Can you please avoid inline styles? See https://github.com/phpmyadmin/phpmyadmin/issues/12262

nijel commented on libraries/classes/DbMultiTableQuery.php:
```diff
@@ -0,0 +1,187 @@
+        $html_output .= '</div>';
+        $html_output .= '<a href="#" title="Remove this column" style="float:right;color:red" class="removeColumn">X</a>';
+        $html_output .= '</fieldset>';
+        return $html_output;
```
It would be much better to use templates than generating HTML this way.

nijel commented on js/db_multi_table_query.js:
```diff
@@ -0,0 +1,228 @@
+        }
+
+        query = 'SELECT ';
+        query += '`' + columns[0][0] + '`.`' + columns[0][1] + '`';
```
What if column name contains backtick? Generally this code is missing escaping.

Achilles-96 replied:
Do we have any in-built escape function?

nijel replied:
Yes, use ``Util::backquote`` (it adds the backticks as well).

ibennetch commented:
Indeed, here's a comment I tried to post over the weekend:

Also, we need to add single quotes (`'`) surrounding the information and escape the text the user can enter:

![image](https://user-images.githubusercontent.com/1443710/28494071-7d34fb04-6ef1-11e7-9f26-84ac8272fc7c.png)

I believe that's as easy as adding the single quotes in your query and using the DBI function  `escapeString`. I suppose one could debate about whether integer values should be surrounded by single quotes (they shouldn't, technically, but MySQL seems to automatically cast and do the right thing here).

Achilles-96 commented on libraries/classes/DbMultiTableQuery.php:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private function getColumnsHTML()
+    {
```
@nijel, do you think this function also better templated?

nijel replied:
I would prefer to use templates for all HTML generating, but I won't enforce this ;-)

ibennetch commented:
Please correct me if I'm wrong, but I believe this still lacks proper escaping; for instance a query with a single quote (such as `d'Arnaud`) fails.

ibennetch commented:
Good work @Achilles-96!

lem9 commented:
Hi all,
in https://github.com/phpmyadmin/phpmyadmin/wiki/GSoC_2017_ideas_list I mentionned "the blocks could be moved left/right with the mouse, because their left-right order represent the order of the generated query". Is there a way to move the blocks in this pull request?

Also, what the point of having the "Query window" link for slider? When this is closed, nothing is left.

nijel commented:
Also the Selenium tests seem to no longer work:

```
1) PmaSeleniumQueryByExampleTest::testQueryByExample

InvalidArgumentException: Element not found.

/home/travis/build/phpmyadmin/phpmyadmin/test/selenium/TestBase.php:390

/home/travis/build/phpmyadmin/phpmyadmin/test/selenium/PmaSeleniumQueryByExampleTest.php:69
```

devenbansod commented:
Hi, I encountered a few issues while using the new interface:
* The `SORT` part of the query don't get generated
* I couldn't figure out a way to specify the sort order between two columns (for example to generate a clause like `ORDER BY article_rating DESC, article_time ASC`) which was present in older interface
* Do we have plans to introduce a better version of the visual builder or was it to be completely replaced by this?
